bindings/wallet-core/src/c.rs

```diff
@@ -385,6 +393,107 @@ pub fn wallet_set_state(wallet: WalletPtr, value: u64, counter: u32) -> Result {
     Result::success()
 }
 
+pub unsafe fn wallet_vote_plan(
+    id: *const u8,
+    payload_type: u8,
+    vote_plan_out: *mut VotePlanPtr,
+) -> Result {
+    if id.is_null() {
+        return Error::invalid_input("id").with(NulPtr).into();
+    }
+
+    if vote_plan_out.is_null() {
+        return Error::invalid_input("vote_plan_out").with(NulPtr).into();
+    }
+
+    let payload_type = match payload_type.try_into() {
+        Ok(payload_type) => payload_type,
+        Err(err) => return Error::invalid_input("payload_type").with(err).into(),
+    };
+
+    let id = std::slice::from_raw_parts(id, crate::vote::VOTE_PLAN_ID_LENGTH);
+    let id = match VotePlanId::try_from(id) {
+        Ok(id) => id,
+        Err(err) => return Error::invalid_input("id").with(err).into(),
+    };
+
+    *vote_plan_out = Box::into_raw(Box::new(VotePlan::new(id, payload_type)));
+
+    Result::success()
+}
+
+pub unsafe fn wallet_vote_proposal(
+    index: u8,
+    num_choices: u8,
+    proposal_out: *mut ProposalPtr,
+) -> Result {
+    if proposal_out.is_null() {
+        return Error::invalid_input("proposal_out").with(NulPtr).into();
+    }
+
+    let options = match VoteOptions::new_length(num_choices) {
+        Ok(options) => options,
+        Err(err) => return Error::invalid_input("num_choices").with(err).into(),
+    };
+
+    *proposal_out = Box::into_raw(Box::new(Proposal::new(index, options)));
+
+    Result::success()
+}
+
+pub unsafe fn wallet_vote_cast(
+    wallet: WalletPtr,
+    settings: SettingsPtr,
+    vote_plan: VotePlanPtr,
+    proposal: ProposalPtr,
+    choice: u8,
+    transaction_out: *mut *const u8,
+    len_out: *mut usize,
+) -> Result {
+    let wallet = if let Some(wallet) = wallet.as_mut() {
+        wallet
+    } else {
+        return Error::invalid_input("wallet").with(NulPtr).into();
+    };
+
+    let settings = if let Some(settings) = settings.as_ref() {
+        settings.clone()
+    } else {
+        return Error::invalid_input("settings").with(NulPtr).into();
+    };
+
+    let vote_plan = if let Some(vote_plan) = vote_plan.as_ref() {
+        vote_plan
+    } else {
+        return Error::invalid_input("vote_plan").with(NulPtr).into();
+    };
+
+    let proposal = if let Some(proposal) = proposal.as_ref() {
+        proposal
+    } else {
+        return Error::invalid_input("proposal").with(NulPtr).into();
+    };
+
+    if transaction_out.is_null() {
+        return Error::invalid_input("transaction_out").with(NulPtr).into();
+    }
+    if len_out.is_null() {
+        return Error::invalid_input("len_out").with(NulPtr).into();
+    }
+
+    let choice = Choice::new(choice);
+
+    let transaction = match wallet.vote(settings, vote_plan, proposal, choice) {
+        Ok(transaction) => transaction,
+        Err(err) => return Error::from(err).into(),
+    };
+
+    *transaction_out = transaction.as_ptr();
```

Comment:
Shouldn't we leak this and delete it later? Or am I missing something?

Reply:
Ah, you are right. Will do that.

Reply:
the `Vec` will be de-scoped and the array will be `free`ed. Need to actually leak it.